chore: Map dateModified to dateCreated

eikek opened the pull request:
When CLI passes a dateModified, it copies dateCreated from the first dataset or plan. We keep dateCreated on each entity as the date the specific entity was created and so map the dateModified to our dateCreated. To get the dateCreated of the original dataset/plan, we can follow the `topMostSameAs` edge.

Issue: #1259

eikek committed:
chore: Map dateModified to dateCreated

When CLI passes a dateModified, it copies dateCreated from the first
dataset or plan. We keep dateCreated on each entity as the date the
specific entity was created and so map the dateModified to our
dateCreated. To get the dateCreated of the original dataset/plan, we
can follow the `topMostSameAs` edge.

eikek committed:
chore: refactoring provenance conversion from cli data

eikek committed:
Match dateModified and map to dateCreated

jachro reviewed (approved):
Looks good in my eyes

eikek committed:
Fix tests by generating correct cli data

Now the cli data and the structures we store in our jena database are
different in encoding provenance. That enforces us now to better
distinguish the target of the encoders and decoders and requires to
"inject" a different provenance encoder into the tests that work with
cli data structures.

eikek committed:
WIP: provide cli encoders for datasets

eikek committed:
WIP: trying to pin down decoding tests

eikek committed:
WIP: more steps

jachro committed:
Merge branch 'development'

jachro committed:
fix: merge with the new Cli model classes

jachro committed:
fix: flaky CliActivitySpec

jachro committed:
feat: CliPublicationEvent added

jachro committed:
fix: adjustments around CliPerson modeling; first DS spec fixed

jachro committed:
refactor: sameAs modeling on the CliDataset

jachro committed:
feat: some more DatasetSpecs converted

jachro committed:
feat: some more DatasetSpecs converted

jachro committed:
feat: some more DatasetSpecs converted

jachro committed:
feat: flaky CliPlanSpec

jachro committed:
chore: missing preambles

jachro committed:
fix: ProjectSpec

jachro committed:
fix: flaky tests

jachro committed:
fix: DatasetPart generator to produce File location Entities only

jachro committed:
fix: EntityBuilderSpec

jachro committed:
fix: EntityBuilderSpec

jachro committed:
refactor: CliCollection -> CliEntityCollection

jachro committed:
refactor: CliEntityCollection -> CliCollectionEntity

jachro committed:
refactor: CliEntity -> CliSingleEntity

jachro committed:
refactor: GenerationEntity -> CliEntity

jachro committed:
refactor: CliConverters split

jachro committed:
feat: cli entities re-modeling

jachro committed:
feat: cli converters continued

jachro committed:
feat: fix all cli model compilation issues

jachro committed:
feat: all cli model converters

jachro committed:
fix: EntityBuilderSpec

jachro committed:
Merge branch 'development'

jachro committed:
refactor: formatting

jachro committed:
Merge branch 'development'

eikek commented on renku-cli-model/src/main/scala/io/renku/cli/model/CliModel.scala:
```diff
@@ -21,7 +21,7 @@ package io.renku.cli.model
 
 /** Marker trait to annotate a class that models an entity from the cli schema. */
-trait CliModel
+trait CliModel extends Product with Serializable
```
Is this really necessary? i.e. the `Serializable`?

jachro replied:
Actually, I think I can take it out

eikek commented on renku-cli-model/src/test/scala/io/renku/cli/model/diffx/CliDiffInstances.scala:
```diff
@@ -24,71 +24,77 @@ import io.renku.graph.model.diffx.ModelTinyTypesDiffInstances
 trait CliDiffInstances extends ModelTinyTypesDiffInstances {
 
-  implicit val cliPersonDiff: Diff[CliPerson] = Diff.derived[CliPerson]
+  implicit lazy val cliPersonDiff: Diff[CliPerson] = Diff.derived[CliPerson]
```
I tried to get the order right: :-) but `lazy` is actually more convenient 👍🏼

jachro replied:
You caught me on naughty things :)

eikek commented on renku-cli-model/src/test/scala/io/renku/cli/model/generators/SoftwareAgentGenerators.scala:
```diff
@@ -0,0 +1,34 @@
+trait SoftwareAgentGenerators {
+
+  def softwareAgentGen: Gen[CliSoftwareAgent] =
+    for {
```
Just wondering: I saw you changed many for comprehensions to start on the `def` line - but this one is on the line below (which is my personal pref 😆 so I like it hehe). Is this perhaps a written formatting rule somewhere?

jachro replied:
I don't think we've got a rule rewriting that anywhere. I somehow have a clashing preference so it might be that when I do some changes in a file where I see your formatting, my auto-manual formatter could kick-off. Sorry...

eikek replied:
No worries! I don't have a strong opinion here and might also mix it sometimes. was genuinely wondering as I looked through

eikek reviewed (commented):
Epic!! 🚀

jachro committed:
refactor: CliModel not to be serializable
